### PR TITLE
[VTAdmin] Migrate to Vite

### PR DESCRIPTION
**Pull request by notfelineit**
## Description
This PR addresses [this issue](https://github.com/vitessio/vitess/issues/12280) and migrates VTAdmin web from create-react-app to vite. Some of the larger changes include:
- Switching to `vitest` for all tests
    - Removes `jest` since `vitest` replaces it
    - Removes repetitive beforeAll/afterEach/afterAll since `vitest` has a global config file 
- Renaming `REACT_APP_*` env variables to `VITE_`
- Changing `process.env` to `import.meta.env`
- Changing protobuf output from `commonjs` to `es6` to better work with vite's defaults
- Configures eslint with vite and fixes some lint errors
- Fixes some prettier errors

## Manual validations
- Ran `./101_initial_cluster.sh` and VTAdmin web loads fine 👍 
- Ran `npm run build` locally and verified build runs locally
- Ran `npm run local` and verified env vars passed are respected
## Related Issue(s)
- **Issue:** https://github.com/vitessio/vitess/issues/12280
- **Web docs PR:** https://github.com/vitessio/website/pull/1437
## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported **No need for backport**
-   [x] Tests were added or are not required **Tests have been modified to use vitest**
-   [x] Did the new or modified tests pass consistently locally and on the CI **Yes**
-   [x] Documentation was added or is not required **Docs are [here](https://github.com/vitessio/website/pull/1437)**

## Deployment Notes
N/A

### CODE REVIEW

**Comment by notfelineit** [comment]
I think we'll have to disable this workflow for this PR: https://github.com/vitessio/vitess/actions/runs/4620726123/workflow?pr=12831 since the PR itself modifies the `make vtadmin_web_proto_types` script.

File: changelog/17.0/17.0.0/summary.md [hunk 1]
```diff
@@ -5,7 +5,11 @@
 - **[Major Changes](#major-changes)**
   - **[Breaking Changes](#breaking-changes)**
     - [Dedicated stats for VTGate Prepare operations](#dedicated-vtgate-prepare-stats)
+<<<<<<< HEAD
+    - [VTAdmin web migrated from create-react-app to vite](#migrated-vtadmin)
+=======
     - [Keyspace name validation in TopoServer](#keyspace-name-validation)
+>>>>>>> bf0af2b7994ce21585052ee4fa76b17039d2b89a
```
Comment (notfelineit) [inline changelog/17.0/17.0.0/summary.md:12]:
```suggestion
    - [VTAdmin web migrated from create-react-app to vite](#migrated-vtadmin)
    - [Keyspace name validation in TopoServer](#keyspace-name-validation)
```

File: web/vtadmin/bin/generate-proto-types.sh [hunk 2]
```diff
@@ -35,7 +35,8 @@ $pbjs_bin \
 	--keep-case \
 	-p "$vitess_proto_root" \
 	-t static-module \
-	-w commonjs \
+  --es6 \
+	-w es6 \
```
Comment (frouioui) [inline web/vtadmin/bin/generate-proto-types.sh:39]:
```suggestion
        --es6 \
	-w es6 \
```

File: web/vtadmin/tests/handlers.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import { graphql, rest } from 'msw'
+
```
Comment (frouioui) [inline web/vtadmin/tests/handlers.ts:2]: We should have a header here

File: web/vtadmin/tests/server.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import { setupServer } from 'msw/node'
+import { handlers } from './handlers'
```
Comment (frouioui) [inline web/vtadmin/tests/server.ts:2]: Here as well

File: web/vtadmin/tests/setup.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import { expect, afterEach, vi } from 'vitest';
+import { cleanup } from '@testing-library/react';
```
Comment (frouioui) [inline web/vtadmin/tests/setup.js:2]: Here too

**Comment by frouioui** [comment]
Left several comments about license header missing, but there are a few other places where the headers are missing

File: changelog/17.0/17.0.0/summary.md [hunk 6]
```diff
@@ -50,6 +51,13 @@ Here is a (condensed) example of stats output:
 }
 ```
 
+#### <a id="migrate-vtadmin"/>VTAdmin web migrated to vite
```
Comment (frouioui) [inline changelog/17.0/17.0.0/summary.md:54]:
```suggestion
#### <a id="migrated-vtadmin"/>VTAdmin web migrated to vite
```

**Review by frouioui — APPROVED** [review]
I am not an expert in VTAdmin so I'll let @ajm188 stamp this. But everything outside of VTAdmin and the few comments I left looks good to me 🙏🏻

File: web/vtadmin/tests/handlers.ts (hunk 3 above)
Comment (notfelineit) [reply web/vtadmin/tests/handlers.ts:2]: Ah thank you! Totally forgot about those!

File: web/vtadmin/tests/handlers.ts [hunk 7]
```diff
@@ -0,0 +1,5 @@
...
+export const handlers = [
+
+]
```
Comment (ajm188) [inline web/vtadmin/tests/handlers.ts:5]: nit: missing a trailing newline here

File: web/vtadmin/tests/server.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+// The big downside to mocking or "faking" APIs like vtadmin is that
+// we end up re-implementing some (or all) of vtadmin-api in our test environment.
+// It is, unfortunately, impossible to completely avoid this kind of duplication
+// unless we solely use e2e tests (which have their own trade-offs).
+//
+// That said, our use of protobufjs to validate and strongly type HTTP responses
+// means our fake is more robust than it would be otherwise. Since we are using
+// the exact same protos in our fake as in our real vtadmin-api server, we're guaranteed
+// to have type parity.
+export const server = setupServer(...handlers)
```
Comment (ajm188) [inline web/vtadmin/tests/server.ts:23]: nit: newline

File: web/vtadmin/tsconfig.node.json [hunk 9]
```diff
@@ -0,0 +1,9 @@
+{
+  "compilerOptions": {
+    "composite": true,
+    "module": "ESNext",
+    "moduleResolution": "Node",
+    "allowSyntheticDefaultImports": true
+  },
+  "include": ["vite.config.ts"]
+}
```
Comment (ajm188) [inline web/vtadmin/tsconfig.node.json:9]: nit: newline

File: web/vtadmin/vite.config.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+      eslint()
+    ],
+    test: {
+      globals: true,
+      setupFiles: './tests/setup.js',
+      environment: 'jsdom',
+    },
+    base: './'
+  };
+});
```
Comment (ajm188) [inline web/vtadmin/vite.config.ts:37]: nit: newline

**Review by ajm188 — APPROVED** [review]
LGTM! Just a couple small things, feel free to fix-and-merge!

**Comment by dbussink** [comment]
> I think we'll have to disable this workflow for this PR: https://github.com/vitessio/vitess/actions/runs/4620726123/workflow?pr=12831 since the PR itself modifies the make vtadmin_web_proto_types script.

That shouldn't be needed if what the new version creates is also committed?

File: web/vtadmin/README.md [hunk 11]
```diff
@@ -14,32 +14,33 @@ Scripts for common and not-so-common tasks. These are always run from the `vites
 |---|---|
 | `npm local` | Start vtadmin-web in development mode on [http://localhost:3000](http://localhost:3000), pointed at a vtadmin-api server running on [http://localhost:14200](http://localhost:14200). This is most useful when running against a [local Vitess cluster](https://vitess.io/docs/get-started/local/). |
 | `npm start` | Start vtadmin-web in development mode on [http://localhost:3000](http://localhost:3000). Additional environment variables can be specified on the command line or in a .env file; see [Environment Variables](#environment-variables). |
-| `npm run test` | Launches the test runner in the interactive watch mode. See the create-react-app documentation about [running tests](https://facebook.github.io/create-react-app/docs/running-tests) for more information. |
+| `npm run test` | Launches the test runner in the interactive watch mode. See the vitest documentation about [running tests](https://vitest.dev/guide/cli.html#vitest-run) for more information. |
```
[the remark this answers is not in the document]
Comment (notfelineit) [reply web/vtadmin/README.md:17]: 😅 yeah it will be, we could add (NOT VITEST!!) afterwards lol

File: web/vtadmin/src/api/http.ts [hunk 12]
```diff
@@ -41,13 +41,14 @@ export const vtfetch = async (endpoint: string, options: RequestInit = {}): Prom
...
 
-        const url = `${env().REACT_APP_VTADMIN_API_ADDRESS}${endpoint}`;
+        const url = `${env().VITE_VTADMIN_API_ADDRESS}${endpoint}`;
         const opts = { ...vtfetchOpts(), ...options };
 
         let response = null;
         try {
             response = await global.fetch(url, opts);
         } catch (error) {
+            console.log(error);
```
[the remark this answers is not in the document]
Comment (notfelineit) [reply web/vtadmin/src/api/http.ts]: It's leftover 😓 removing!

**Comment by notfelineit** [comment]
@dbussink we wouldn't need to disable it if the workflow uses this branch's `generate-proto-types.sh` file 🤔 wasn't sure if that was the case or not

**Comment by dbussink** [comment]
> @dbussink we wouldn't need to disable it if the workflow uses this branch's `generate-proto-types.sh` file 🤔 wasn't sure if that was the case or not.

It would be, afaik all jobs run against the current branch including this one so it would generate the same code.

**Comment by notfelineit** [comment]
@dbussink not sure why the CI is failing then 🤔 even after pulling latest from main + running `make vtadmin_web_proto_types` the proto files don't change - I think they're already up-to-date.

**Comment by notfelineit** [comment]
@dbussink thank you! Updating now and checking the diff. Hoping it shows!

File: .gitattributes [hunk 13]
```diff
@@ -1,6 +1,3 @@
-web/vtadmin/src/proto/** linguist-generated=true
-web/vtadmin/src/proto/*.js -diff
-web/vtadmin/src/proto/*.ts -diff
```
Comment (dbussink) [inline .gitattributes:3]: @notfelineit Should we revert this now that the issue was found?

File: web/vtadmin/package.json [hunk 14]
```diff
@@ -33,23 +31,22 @@
     "react-toastify": "^8.1.0",
```
Comment (dbussink) [inline web/vtadmin/package.json:31]: @notfelineit I'm seeing also a following message:

```
npm WARN deprecated react-flow-renderer@10.3.17: react-flow-renderer has been renamed to reactflow, please use this package from now on https://reactflow.dev/docs/guides/migrate-to-v11/
```

Should we also update all dependencies here in general? Or do that in a separate PR then?

**Review by dbussink — APPROVED** [review]

File: web/vtadmin/package.json (hunk 14 above)
Comment (notfelineit) [reply web/vtadmin/package.json:31]: I'm thinking we can do that in a separate PR! Should be easy to upgrade now we're on Vite 😄

File: .gitattributes (hunk 13 above)
Comment (notfelineit) [reply .gitattributes]: Yes! 😄
